File: csharp/nucleotide-count/DNA.cs
using System;
using System.Collections.Generic;

public class DNA
{
    private readonly Dictionary<char, int> nucleotideCounts = new Dictionary<char, int> { { 'A', 0 }, { 'T', 0 }, { 'C', 0 }, { 'G', 0 } };

    public DNA(string strand)
    {
        foreach (var nucleotide in strand)
        {
            this.nucleotideCounts[nucleotide]++;
        }
    }

    public Dictionary<char, int> NucleotideCounts
    {
        get
        {
            return this.nucleotideCounts;
        }
    }

    public int Count(char c)
    {
        if (this.nucleotideCounts.ContainsKey(c))
        {
            return this.nucleotideCounts[c];
        }

        throw new InvalidNucleotideException();
    }
}

public class InvalidNucleotideException : Exception
{
}

File: haskell/nucleotide-count/DNA.hs
module DNA (count, nucleotideCounts) where

import Data.Map (Map, fromList)

validNucleotides :: String
validNucleotides = "ATCG"

count :: Char -> String -> Int
count nucleotide strand 
    | isValid = length $ filter (== nucleotide) strand
    | otherwise = error $ "invalid nucleotide " ++ show(nucleotide) where
        isValid = elem nucleotide validNucleotides

nucleotideCounts :: String -> Map Char Int
nucleotideCounts strand = fromList $ map (\x -> (x, count x strand)) validNucleotides